### DISCUSSION TITLE
Resync hat bei mir keinen Effekt

### DISCUSSION THREAD

**Discussion by alurio** [audiamus/BookLibConnect#22] [General]
Hallo ! 

Ich habe auf Version 0.4 upgedatet und den Resync durchgeführt.  Die Hoffnung war, dass das Buch  Der Herr des Wüstenplaneten, dass ich vor einiger Zeit zurückgegeben habe verschwindet. Dafür ist die Funktion ja auch da. Neue Bücher aufnehmen alte aus der Bibliothek entfernen. Bei mir ändert sich aber rein gar nichts. Das Buch steht auch nach mehreren Resyncs in der Liste. 

Grüße

**Comment by audiamus** [comment]
Der Mechanismus dahinter ist einigermaßen raffiniert. Ich habe versucht, mehrere Szenarien abzubilden. Es soll schon vorgekommen sein, dass Bücher, die man gekauft und auch nicht zurückgegeben hat, trotzdem aus der Bibliothek verschwinden können, z.B. weil sich ein Lizenzstreit entwickelt hat. Es soll auch Bücher geben, die nur zeitweise verschwunden sind, nach Reklamation bei Audible aber wieder hergestellt werden. 

Was also in Book Lib Connect passiert ist, dass auf Server-Seite fehlende Bücher in der lokalen Datenbank nicht wirklich gelöscht werden, sonder nur als gelöscht markiert. Was man in der Liste zu sehen bekommt, hängt im weiteren aber vom letzten Zustand ab, und auch noch, ob der verifizierbar ist und ob es ein einteiliges oder mehrteiliges Buch ist und wie die zugehörige Einstellung steht.

Wurde das Buch heruntergeladen und konvertiert/exportiert und ist die zum letzten Status passende Datei vorhanden, dann ändert sich in der Anzeige in der Tat nichts. Wurde das Buch aber noch nicht heruntergeladen, oder sind die Zwischendateien nicht mehr vorhanden, dann wird der Status auf unbekannt gesetzt und das Symbol zum rot durchgestrichenen Globus.   

Wenn ein nur zeitweilig gelöschtes Buch dann später wieder auftaucht, wird mit Resync kein neuer Eintrag in der lokalen Datenbank angelegt, sondern der bisherige auf "nicht gelöscht" rückgesetzt.

Kann man sich mit einem SQLite-Tool anschauen, in der Tabelle "Books". Dort gibt es eine Spalte "Deleted". Der Wert ist im Grundzustand <null>, also nicht besetzt. wird bei nicht-mehr-Vorhandensein zu 1 und bei Wiederauftauchen zu 0. 

Der Schritt von 0.3 auf 0.4 ist ist mit einer Änderung dieser Tabelle verbunden. Es gab bis her schon eine Spalte "Removed", die eigentlich für die Löschmarkierung angedacht war. Die war aber nicht "nullable", und damit für die dich eher seltene Wiedereingliederung eines Buches weniger geeignet, weil ein Filterkriterium <null> fehlte.

Bei ersten Start jeder neuen Version von Book Lib Connect werden Änderungen an der Datenbankstruktur automatisch durchgeführt. Man kann auch nicht zurück. Versuchte man, die Version 0.3 zu starten, obwohl bereits vorher 0.4 gelaufen ist, wird das scheitern, weil die Datenbankstruktur nicht mehr passt. Ich überlege, zukünftig eine Backup-Kopie machen zu lassen, bevor die Datenbank umgebaut wird.

**Comment by alurio** [comment]
Hallo !
Ich hab das natürlich ursprünglich mal runtergeladen und konvertiert, fands dann öde und hab es dann  eben zurückgegeben. Status ist jetzt ein rot durchgestrichener Globus. 
Dann hat das ja seine Richtigkeit. und der Titel  verschwindet nicht mehr ganz, 

Viele Grüße
